### ISSUE TITLE
What is the best way to extend the student to multistep?

### ISSUE THREAD

**Issue by sravansripada** [mingyuanzhou/SiD-LSG#5]
Currently, the model shows impressive results for a single step generation. I was wondering if you have some pointers for distilling a multi step student - lets say more than 10 steps.

**Comment by mingyuanzhou** [comment]
We did not advertise the ability to train and eval with multisteps, but the code can already do it:

Simply set num_steps = num_steps_eval = 10 in sid_training_loop.py.

In fact, you can already use the one-step generator to do multistep generation by simply setting num_steps_eval >1, e.g., num_steps_eval = 4.

I am working on an updated version of guided SiD, which will significantly improve its speed (using AMP), scalability (using FSDP), and performance (adding diffusion GAN). Please stay tuned.

**Comment by sravansripada** [comment]
Thank you . Super excited for the future releases. Regarding multistep, I see in the following code that we run forward pass num_step times on the generator. I am just worried that this will slow down training time especially if the student steps are > 10? Was wondering if you had any insights on this?

```
 if train_sampler:
        D_x = torch.zeros_like(latents).to(latents.device)
        step_indices = [torch.tensor(0).to(latents.device)]  # Initial step
        for i in range(num_steps):
            noise = latents if i == 0 else torch.randn_like(latents).to(latents.device)
            init_timesteps_i = (init_timesteps*(1-i/num_steps)).to(torch.long)
            latents = noise_scheduler.add_noise(D_x, noise, init_timesteps_i).to(torch.float32)
            latent_model_input = noise_scheduler.scale_model_input(latents, init_timesteps_i) 
            noise_pred = unet(latent_model_input.to(dtype), init_timesteps_i, encoder_hidden_states=text_embeddings).sample.to(torch.float32)
            D_x  = noise_scheduler.step(noise_pred, init_timesteps_i[0], latents,return_dict=True).pred_original_sample.to(torch.float32)  
```

**Comment by mingyuanzhou** [comment]
Using the generator as G(G(G...(z))) will slow down training because it increases both the computational cost of generating fake images and the cost of computing gradients for G. Additionally, as the number of steps increases, it is advisable to reduce G's learning rate to maintain stability. While simply adjusting the learning rate can allow the code to run, more sophisticated algorithmic designs may be necessary to fully optimize performance, given the inherent challenges of training deeper models.

**Comment by mingyuanzhou** [comment]
I am closing the issue for now. Please stay tuned for our upcoming code update, which will include both a simple approach and a more advanced optimization method for multistep generation.
